torch/distributed/_sharded_tensor/ops/linear.py

```diff
@@ -218,15 +268,144 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
+        local_shard_t: row-wise shared local weight used for lookup.
+        bias: bias term of linear op.
+        pg: process group.
+
+    Returns:
+        A :class:`_PartialTensor` object which stores the partial local result.
+    """
+    results = []
+    local_shard = input.local_shards()[0].tensor
+    indices = [0] * world_size
+    reaggrance_partial = False
+    for idx, placement in enumerate(input._sharding_spec.placements):
+        indices[placement.rank()] = idx
+        if idx != placement.rank():
+            reaggrance_partial = True
+
+    for tensor in torch.tensor_split(local_shard, world_size):
+        results.append(
+            tensor.matmul(local_shard_t) + _BiasTensorPartial.apply(world_size, bias)
+        )
+    if reaggrance_partial:
+        results = [results[idx] for idx in indices]
+
+    # Return the partial local result.
+    return _PartialTensor(torch.cat(results), pg)
 
-    # Gather all the results appropriately.
-    local_result = torch.empty_like(results[rank])
-    local_result = reduce_scatter(local_result, results, group=pg)
 
-    # Return the appropriate local result.
-    return local_result + bias
+def _init_sharded_tensor_from_local_result(
+    sharded_tensor,
+    local_result,
+    tensor_shard_dim,
+    result_shard_dim,
+    world_size,
+    pg,
+):
+    """
+    Given a sharded tensor and local_result from an op on top of it. We want
+    to create a new sharded tensor from the local_result so that the the next
+    op can be performed on the basis of the new sharded tensor. This can seen
+    as the last step of the first phase of the Megatron-LM style model(tensor)
+    parallelism.
+
+    Args:
+        sharded_tensor: Sharded tensor which the op was performed on.
+        local_result: A tensor which is from the op performed on the local_shard of
+            the sharded_tensor.
+        tensor_shard_dim: Dim which the tensor is sharded on.
+        result_shard_dim: Dim which the new sharded tensor will be sharded on.
+        world_size: number of ranks.
+        pg (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+
+    Return:
+        A :class:`ShardedTensor` object which filled with local intermediate results.
+    """
+    sharded_weight_metadata = copy.deepcopy(sharded_tensor.local_shards()[0].metadata)
+    current_offsets = [0] * len(local_result.size())
+    current_offsets[result_shard_dim] = sharded_weight_metadata.shard_offsets[
+        tensor_shard_dim
+    ]
+    global_size = list(local_result.size())
+    global_size[result_shard_dim] = sharded_tensor.size(tensor_shard_dim)
+    local_shard_metadata = ShardMetadata(
+        shard_offsets=current_offsets,
+        shard_sizes=list(local_result.size()),
+        placement=sharded_weight_metadata.placement,
+    )
+    local_shards = [Shard(local_result, local_shard_metadata)]
+    new_st = ShardedTensor._init_from_local_shards(
```

Comment:
This has additional allgather overhead as mentioned earlier: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/api.py#L351